gpflow/posterior.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2016-2020 The GPflow Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple
+
+import numpy as np
+import tensorflow as tf
+
+from . import covariances, kernels
+from .base import Module, Parameter
+from .conditionals.util import expand_independent_outputs, mix_latent_gp
+from .config import default_float, default_jitter
+from .models.model import MeanAndVariance
+
+
+class DiagNormal(Module):
+    def __init__(self, q_mu, q_sqrt):
+        self.q_mu = q_mu  # [M, L]
+        self.q_sqrt = q_sqrt  # [M, L]
+
+
+class MvnNormal(Module):
+    def __init__(self, q_mu, q_sqrt):
+        self.q_mu = q_mu  # [M, L]
+        self.q_sqrt = q_sqrt  # [L, M, M], lower-triangular
+
+
+class Posterior(Module):
+    def __init__(self, kernel, iv, q_mu, q_sqrt, whiten=True, mean_function=None):
+        self.iv = iv
+        self.kernel = kernel
+        self.mean_function = mean_function
+        self.whiten = whiten
+        if len(q_sqrt.shape) == 2:  # q_diag
+            self.q_dist = DiagNormal(q_mu, q_sqrt)
+        else:
+            self.q_dist = MvnNormal(q_mu, q_sqrt)
+
+        self.update_cache()  # populates or updates self.alpha and self.Qinv
+
+    def _precompute(self):
+        Kuu = covariances.Kuu(self.iv, self.kernel, jitter=default_jitter())  # [(R), M, M]
+        q_mu = self.q_dist.q_mu
+
+        if Kuu.shape.ndims == 4:
+            ML = tf.cast(tf.sqrt(tf.cast(tf.reduce_prod(tf.shape(Kuu)), default_float())), tf.int32)
+            Kuu = tf.reshape(Kuu, [ML, ML])
+        if Kuu.shape.ndims == 3:
+            q_mu = tf.linalg.adjoint(self.q_dist.q_mu)[..., None]  # [..., R, M, 1]
+        L = tf.linalg.cholesky(Kuu)
+
+        if not self.whiten:
+            # alpha = Kuu⁻¹ q_mu
+            alpha = tf.linalg.cholesky_solve(L, q_mu)
+        else:
+            # alpha = L⁻T q_mu
+            alpha = tf.linalg.triangular_solve(L, q_mu, adjoint=True)
+        # predictive mean = Kfu alpha
+        # predictive variance = Kff - Kfu Qinv Kuf
+        # S = q_sqrt q_sqrtT
+        if not self.whiten:
+            # Qinv = Kuu⁻¹ - Kuu⁻¹ S Kuu⁻¹
+            #      = Kuu⁻¹ - L⁻T L⁻¹ S L⁻T L⁻¹
+            #      = L⁻T (I - L⁻¹ S L⁻T) L⁻¹
+            #      = L⁻T B L⁻¹
+            if isinstance(self.q_dist, DiagNormal):
```

Comment:
you could make a method in `DiagNormal` and `MvnNormal` that returns you the correct q_sqrt parametrisation?